### ISSUE TITLE
Connection dropping out

### ISSUE THREAD

**Issue by jhs** [jhs/node-fastcgi-parser#1]
The problem is, that I can nto get file uploads working - well, not stable.

I have brought it to where I can upload files and - sometimes, and that is the very problem… - have them handed to PHP. But as mentioned, only _sometimes_. The greater the file size becomes, the greater is the problem and the smaller the possibility that the upload will pass through correctly. PHP emits the „Partial Upload“ error, which is emitted when it can not find the end of the file section, separated by the boundary. So going from this, sometimes that very section is not completely handed to PHP…and I just can’t figure out what is being wrong.

You can see my current implementation in [1]. the req.rawBody property is created in [2] and neccessary, since a later middleware hogs up the request stream and will therefore take away the events required for this module - therefore, the raw body is saved back before this happens.

Do you have any idea how I can make sure that the uploads are stable and don’T drop out from time to time?

**Comment by IngwiePhoenix** [comment]
I tried my luck again, but the issues remain. Here are the sources to have a look into my code again:

1: https://github.com/IngwiePhoenix/connect-yii/blob/master/node-php.js#L364-409
2: http://git.ingwie.me/ingwie/bird3/blob/master/app.js#L58-72

**Comment by IngwiePhoenix** [comment]
I also just uploaded results of a test that I used to test upload stability. Tests above 1MB looked way more unstable.

https://gist.github.com/IngwiePhoenix/97c3a81c55e1775fc664

**Comment by jhs** [comment]
Thanks very much! I will attempt to replicate your issues. Do I need to have PHP installed to test this?

**Comment by IngwiePhoenix** [comment]
Yes. The module talks to php via a TCP socket - in my case I use `php-fpm`.

Here is a tiny nodejs script that reproduces a small version of my built-in webserver that runs PHP:

``` javascript
var app = require("express")();
var php = require("connect-yii");
var st = require("express-static");
var http_port = 8081;

app.use(function(req,res,next){
    console.log(req.method+" > "+req.url);
    return next();
});

// Re-Create the req.rawBody
app.use("/", function(req,res,next){
    console.log("o3o");
    if(req.method == "POST") {
        console.log("Starting to save buffer...");
        req._rawBodyParts = [];
        req._rawBodyPartsLength = 0;
        req.rawBody = new Buffer(0);
        req.on("data", function(ch){
            req._rawBodyPartsLength++;
            req._rawBodyParts.push(ch);
        });
        req.on("end", function(){
            console.log("Done.");
            req.rawBody = Buffer.concat(req._rawBodyParts);
            next();
        });
    } else return next();
});

app.use("/", php({
    root: process.cwd(),
    index: "index.php",
    serverName: "localhost",
    serverHost: "127.0.0.1",
    serverPort: http_port
}));

app.use("/", st(process.cwd()));

app.listen(http_port, function(){
    console.log("-- Started on localhost:"+http_port);
});
```

To make it run:

```
$ npm install express express-static connect-yii
$ node test.js
```
